# 07 MongoDB Atlas Development

## NorthStar Urban Mobility and Logistics

This notebook demonstrates the development of a NoSQL database solution using **MongoDB Atlas** for the NorthStar Urban Mobility and Logistics case study.

The notebook shows how operational delivery data can be stored in a document-oriented database, how collections can be created from structured datasets, and how MongoDB queries can be used to retrieve useful business information.

### Main objectives
- Connect Python to MongoDB Atlas
- Load CSV datasets into pandas DataFrames
- Convert relational-style data into MongoDB-compatible documents
- Create MongoDB collections for deliveries, drivers, vehicles, hubs, and customers
- Insert records into MongoDB Atlas
- Run example queries to support operational analysis
- Demonstrate how NoSQL supports flexible and scalable logistics data management

In [ ]:
# Install the MongoDB Python driver required for this notebook
!pip install pymongo

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 33.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 331.1/331.1 kB 14.7 MB/s eta 0:00:00


In [ ]:
# Import the libraries required for MongoDB development and data handling

import pandas as pd
from pymongo import MongoClient
from pprint import pprint

print("Libraries imported successfully.")

Libraries imported successfully.


## 1. Importing the Operational Datasets

The first stage of the MongoDB development process is to load the structured CSV datasets into Python. These datasets represent the main operational entities used by NorthStar, including deliveries, drivers, vehicles, hubs, and customers.

Although the original files are stored in a tabular format, they will later be converted into JSON-like documents so they can be inserted into MongoDB collections.

In [ ]:
# Load the operational datasets directly from the GitHub repository

base_url = "https://raw.githubusercontent.com/PawanKumar1216/northstar-databases-analytics-assignment/main/data/"

deliveries = pd.read_csv(base_url + "deliveries.csv")
drivers = pd.read_csv(base_url + "drivers.csv")
vehicles = pd.read_csv(base_url + "vehicles.csv")
hubs = pd.read_csv(base_url + "hubs.csv")
customers = pd.read_csv(base_url + "customers.csv")

print("Datasets loaded successfully.")
print("Deliveries:", deliveries.shape)
print("Drivers:", drivers.shape)
print("Vehicles:", vehicles.shape)
print("Hubs:", hubs.shape)
print("Customers:", customers.shape)

Datasets loaded successfully.
Deliveries: (950, 13)
Drivers: (170, 8)
Vehicles: (120, 8)
Hubs: (8, 5)
Customers: (650, 9)


## 2. Initial Dataset Inspection

Before inserting the data into MongoDB, the datasets are inspected to confirm their structure, column names, and example records. This helps verify that the imported data is suitable for conversion into MongoDB documents.

In [ ]:
# Display the first few records from each dataset to inspect their structure

print("DELIVERIES DATASET")
display(deliveries.head())

print("\nDRIVERS DATASET")
display(drivers.head())

print("\nVEHICLES DATASET")
display(vehicles.head())

print("\nHUBS DATASET")
display(hubs.head())

print("\nCUSTOMERS DATASET")
display(customers.head())

DELIVERIES DATASET


,delivery_id,order_id,driver_id,vehicle_id,hub_id,dispatch_time,delivery_completed_at,delivery_status,route_distance_km,manual_route_override_count,proof_of_completion_missing,customer_rating_post_delivery,fuel_or_charge_cost
0,DL00001,O00938,D004,V056,H05,2024-06-18 10:57:00,2024-06-19 09:05:59.904311,Failed,17.26,1,0,3.07,12.05
1,DL00002,O00004,D138,V007,H02,2025-01-11 18:45:00,2025-01-11 17:39:00.000000,OnTime,10.34,1,0,5.00,13.41
2,DL00003,O00639,D006,V049,H02,2025-06-02 20:39:00,2025-06-02 21:45:32.366770,OnTime,7.92,0,0,4.98,8.51
3,DL00004,O00313,D116,V055,H02,2024-03-08 23:31:00,2024-03-09 23:30:08.103702,Delayed,16.42,0,0,4.18,13.62
4,DL00005,O00844,D108,V034,H01,2025-09-21 11:43:00,2025-09-21 15:45:34.131056,OnTime,14.52,1,0,4.18,9.22



DRIVERS DATASET


,driver_id,base_zone,employment_type,years_experience,training_score,driver_rating,shift_preference,active_flag
0,D001,AIRPORT,FullTime,8,67.8,3.54,Morning,1
1,D002,Central,FullTime,4,42.4,3.94,Evening,1
2,D003,AIRPORT,FullTime,11,96.5,5.00,Evening,1
3,D004,Airport,PartTime,13,88.9,4.75,Morning,1
4,D005,north,FullTime,3,69.7,4.14,Morning,1



VEHICLES DATASET


,vehicle_id,vehicle_type,assigned_zone,commission_date,battery_health_pct,odometer_km,maintenance_status,telematics_version
0,V001,EV,North,2024-12-28 23:48:00,71.8,56928,Active,v2.2
1,V002,EV,AIRPORT,2024-04-21 16:14:00,67.9,159368,InRepair,v2.2
2,V003,CargoVan,north,2025-11-24 23:59:00,91.7,219359,Active,v2.1
3,V004,Hybrid,RiverSide,2024-06-07 13:21:00,NaN,36310,Active,v2.2
4,V005,CargoVan,West,2025-11-15 11:08:00,58.6,146638,Active,v2.2



HUBS DATASET


,hub_id,hub_name,zone,hub_type,capacity_score
0,H01,North Exchange,North,Dispatch,82
1,H02,South Link,South,Dispatch,78
2,H03,East Dock,East,Warehouse,74
3,H04,West Gate,West,Dispatch,69
4,H05,Central Core,Central,Control,88



CUSTOMERS DATASET


,customer_id,age,home_zone,customer_type,signup_date,loyalty_score,app_engagement_score,preferred_channel,account_status
0,C0001,26,North,SME,2024-11-27 04:25:00,44.9,69.2,App,Active
1,C0002,61,AIRPORT,Consumer,2025-10-28 01:04:00,55.4,66.6,App,Active
2,C0003,66,East,Consumer,2025-07-02 03:23:00,75.9,33.8,NaN,Active
3,C0004,75,CENTRAL,Consumer,2025-08-19 01:58:00,32.5,33.0,App,Active
4,C0005,26,Riverside,Consumer,2025-06-03 06:02:00,55.9,100.0,Web,Active


## 3. Data Preparation for MongoDB

MongoDB stores data as flexible JSON-like documents rather than traditional rows and tables. Before inserting the datasets into MongoDB Atlas, each pandas DataFrame must be converted into a list of Python dictionaries, where each dictionary represents one document in a collection.

Missing values are also converted from `NaN` to `None` so they can be stored correctly as `null` values in MongoDB.

In [ ]:
# Convert pandas DataFrames into MongoDB-compatible documents

def dataframe_to_documents(df):
    """
    Convert a pandas DataFrame into a list of dictionaries suitable for MongoDB.
    Missing values are converted from NaN to None.
    """
    return df.where(pd.notnull(df), None).to_dict(orient="records")

delivery_documents = dataframe_to_documents(deliveries)
driver_documents = dataframe_to_documents(drivers)
vehicle_documents = dataframe_to_documents(vehicles)
hub_documents = dataframe_to_documents(hubs)
customer_documents = dataframe_to_documents(customers)

print("Datasets converted into MongoDB-compatible documents.")
print("Delivery documents:", len(delivery_documents))
print("Driver documents:", len(driver_documents))
print("Vehicle documents:", len(vehicle_documents))
print("Hub documents:", len(hub_documents))
print("Customer documents:", len(customer_documents))

Datasets converted into MongoDB-compatible documents.
Delivery documents: 950
Driver documents: 170
Vehicle documents: 120
Hub documents: 8
Customer documents: 650


In [ ]:
# Display one example MongoDB-style document from the deliveries dataset

print("Example delivery document:")
pprint(delivery_documents[0])

Example delivery document:
{'customer_rating_post_delivery': 3.07,
 'delivery_completed_at': '2024-06-19 09:05:59.904311',
 'delivery_id': 'DL00001',
 'delivery_status': 'Failed',
 'dispatch_time': '2024-06-18 10:57:00',
 'driver_id': 'D004',
 'fuel_or_charge_cost': 12.05,
 'hub_id': 'H05',
 'manual_route_override_count': 1,
 'order_id': 'O00938',
 'proof_of_completion_missing': 0,
 'route_distance_km': 17.26,
 'vehicle_id': 'V056'}


## 4. Connecting to MongoDB Atlas

MongoDB Atlas is used as the cloud-hosted NoSQL database platform for this project. A secure connection string is required to connect Python with the Atlas cluster.

For security reasons, the real password should not be published in the notebook or GitHub repository. The connection string is therefore entered privately when the notebook is run.

In [ ]:
from getpass import getpass
from urllib.parse import quote_plus

username = "northstar_user"
password = quote_plus(getpass("Enter your MongoDB Atlas password: "))

mongo_uri = f"mongodb+srv://{username}:{password}@cluster0.uifgv62.mongodb.net/?appName=Cluster0"

print("MongoDB connection details entered securely.")

Enter your MongoDB Atlas password: ··········
MongoDB connection details entered securely.


In [ ]:
# Connect to MongoDB Atlas and create/select the project database

client = MongoClient(mongo_uri)

# Test the connection
client.admin.command("ping")

# Select the database used for this project
db = client["northstar_logistics"]

print("Successfully connected to MongoDB Atlas.")
print("Database selected: northstar_logistics")

Successfully connected to MongoDB Atlas.
Database selected: northstar_logistics


## 5. Creating MongoDB Collections

In MongoDB, related documents are stored inside collections. For this project, separate collections are created for deliveries, drivers, vehicles, hubs, and customers.

Before inserting the current datasets, any existing collections with the same names are cleared. This prevents duplicate records from being created if the notebook is run more than once.

In [ ]:
# Create/select the MongoDB collections used in the project

deliveries_collection = db["deliveries"]
drivers_collection = db["drivers"]
vehicles_collection = db["vehicles"]
hubs_collection = db["hubs"]
customers_collection = db["customers"]

# Clear existing records to avoid duplicates when the notebook is re-run
deliveries_collection.delete_many({})
drivers_collection.delete_many({})
vehicles_collection.delete_many({})
hubs_collection.delete_many({})
customers_collection.delete_many({})

print("MongoDB collections created and cleared successfully.")

MongoDB collections created and cleared successfully.


In [ ]:
# Insert the prepared documents into their corresponding MongoDB collections

deliveries_collection.insert_many(delivery_documents)
drivers_collection.insert_many(driver_documents)
vehicles_collection.insert_many(vehicle_documents)
hubs_collection.insert_many(hub_documents)
customers_collection.insert_many(customer_documents)

print("Documents inserted into MongoDB Atlas successfully.")

Documents inserted into MongoDB Atlas successfully.


In [ ]:
# Verify the number of documents stored in each MongoDB collection

print("Documents stored in MongoDB Atlas:")
print("Deliveries:", deliveries_collection.count_documents({}))
print("Drivers:", drivers_collection.count_documents({}))
print("Vehicles:", vehicles_collection.count_documents({}))
print("Hubs:", hubs_collection.count_documents({}))
print("Customers:", customers_collection.count_documents({}))

Documents stored in MongoDB Atlas:
Deliveries: 950
Drivers: 170
Vehicles: 120
Hubs: 8
Customers: 650


## 6. Verifying the Stored MongoDB Documents

After inserting the records, a sample document is retrieved directly from MongoDB Atlas. This confirms that the data has been successfully stored in document format inside the NoSQL database.

In [ ]:
# Retrieve and display one delivery document directly from MongoDB Atlas

sample_delivery = deliveries_collection.find_one(
    {},
    {"_id": 0}
)

print("Sample delivery document retrieved from MongoDB Atlas:")
pprint(sample_delivery)

Sample delivery document retrieved from MongoDB Atlas:
{'customer_rating_post_delivery': 3.07,
 'delivery_completed_at': '2024-06-19 09:05:59.904311',
 'delivery_id': 'DL00001',
 'delivery_status': 'Failed',
 'dispatch_time': '2024-06-18 10:57:00',
 'driver_id': 'D004',
 'fuel_or_charge_cost': 12.05,
 'hub_id': 'H05',
 'manual_route_override_count': 1,
 'order_id': 'O00938',
 'proof_of_completion_missing': 0,
 'route_distance_km': 17.26,
 'vehicle_id': 'V056'}


## 7. Basic MongoDB Queries

The following queries demonstrate how MongoDB can retrieve useful operational information from the stored delivery documents. These examples show the practical use of document-based querying for identifying delayed deliveries, failed deliveries, and records with missing proof of completion.

In [ ]:
# Query 1: Retrieve delayed deliveries

delayed_deliveries = list(
    deliveries_collection.find(
        {"delivery_status": "Delayed"},
        {"_id": 0, "delivery_id": 1, "hub_id": 1, "driver_id": 1, "route_distance_km": 1}
    ).limit(5)
)

print("First five delayed deliveries:")
for delivery in delayed_deliveries:
    pprint(delivery)

First five delayed deliveries:
{'delivery_id': 'DL00004',
 'driver_id': 'D116',
 'hub_id': 'H02',
 'route_distance_km': 16.42}
{'delivery_id': 'DL00006',
 'driver_id': 'D037',
 'hub_id': 'H03',
 'route_distance_km': 13.84}
{'delivery_id': 'DL00007',
 'driver_id': 'D151',
 'hub_id': 'H07',
 'route_distance_km': 32.72}
{'delivery_id': 'DL00017',
 'driver_id': 'D002',
 'hub_id': 'H05',
 'route_distance_km': 20.79}
{'delivery_id': 'DL00028',
 'driver_id': 'D165',
 'hub_id': 'H08',
 'route_distance_km': 15.53}


In [ ]:
# Query 2: Count deliveries by delivery status

status_counts = deliveries_collection.aggregate([
    {
        "$group": {
            "_id": "$delivery_status",
            "total_deliveries": {"$sum": 1}
        }
    },
    {
        "$sort": {"total_deliveries": -1}
    }
])

print("Delivery counts by status:")
for status in status_counts:
    pprint(status)

Delivery counts by status:
{'_id': 'OnTime', 'total_deliveries': 616}
{'_id': 'Delayed', 'total_deliveries': 202}
{'_id': 'Failed', 'total_deliveries': 132}


In [ ]:
# Query 3: Retrieve deliveries where proof of completion is missing

missing_proof_deliveries = list(
    deliveries_collection.find(
        {"proof_of_completion_missing": 1},
        {
            "_id": 0,
            "delivery_id": 1,
            "delivery_status": 1,
            "hub_id": 1,
            "proof_of_completion_missing": 1
        }
    ).limit(5)
)

print("First five deliveries with missing proof of completion:")
for delivery in missing_proof_deliveries:
    pprint(delivery)

First five deliveries with missing proof of completion:
{'delivery_id': 'DL00033',
 'delivery_status': 'Failed',
 'hub_id': 'H06',
 'proof_of_completion_missing': 1}
{'delivery_id': 'DL00041',
 'delivery_status': 'Failed',
 'hub_id': 'H08',
 'proof_of_completion_missing': 1}
{'delivery_id': 'DL00052',
 'delivery_status': 'Delayed',
 'hub_id': 'H08',
 'proof_of_completion_missing': 1}
{'delivery_id': 'DL00055',
 'delivery_status': 'Delayed',
 'hub_id': 'H04',
 'proof_of_completion_missing': 1}
{'delivery_id': 'DL00060',
 'delivery_status': 'Delayed',
 'hub_id': 'H02',
 'proof_of_completion_missing': 1}


In [ ]:
# Query 4: Identify deliveries with high manual route override counts

high_override_deliveries = list(
    deliveries_collection.find(
        {"manual_route_override_count": {"$gte": 3}},
        {
            "_id": 0,
            "delivery_id": 1,
            "hub_id": 1,
            "delivery_status": 1,
            "manual_route_override_count": 1,
            "customer_rating_post_delivery": 1
        }
    ).sort("manual_route_override_count", -1).limit(5)
)

print("Top five deliveries with high manual route override counts:")
for delivery in high_override_deliveries:
    pprint(delivery)

Top five deliveries with high manual route override counts:
{'customer_rating_post_delivery': 4.01,
 'delivery_id': 'DL00473',
 'delivery_status': 'OnTime',
 'hub_id': 'H07',
 'manual_route_override_count': 7}
{'customer_rating_post_delivery': 4.94,
 'delivery_id': 'DL00672',
 'delivery_status': 'OnTime',
 'hub_id': 'H02',
 'manual_route_override_count': 5}
{'customer_rating_post_delivery': 3.73,
 'delivery_id': 'DL00374',
 'delivery_status': 'OnTime',
 'hub_id': 'H01',
 'manual_route_override_count': 5}
{'customer_rating_post_delivery': 2.78,
 'delivery_id': 'DL00055',
 'delivery_status': 'Delayed',
 'hub_id': 'H04',
 'manual_route_override_count': 5}
{'customer_rating_post_delivery': 3.85,
 'delivery_id': 'DL00085',
 'delivery_status': 'OnTime',
 'hub_id': 'H06',
 'manual_route_override_count': 5}


## 8. Aggregation Pipeline Analysis

MongoDB aggregation pipelines are used to perform more advanced analysis directly inside the database. The following examples summarise delivery performance by hub and compare the average customer rating across different delivery outcomes.

In [ ]:
# Convert pandas DataFrames into MongoDB-compatible documents

def dataframe_to_documents(df):
    """
    Convert a pandas DataFrame into a list of dictionaries suitable for MongoDB.
    Missing values are converted from NaN to None so MongoDB stores them as null.
    """
    return df.astype(object).where(pd.notnull(df), None).to_dict(orient="records")

delivery_documents = dataframe_to_documents(deliveries)
driver_documents = dataframe_to_documents(drivers)
vehicle_documents = dataframe_to_documents(vehicles)
hub_documents = dataframe_to_documents(hubs)
customer_documents = dataframe_to_documents(customers)

print("Datasets converted into MongoDB-compatible documents.")
print("Delivery documents:", len(delivery_documents))
print("Driver documents:", len(driver_documents))
print("Vehicle documents:", len(vehicle_documents))
print("Hub documents:", len(hub_documents))
print("Customer documents:", len(customer_documents))

Datasets converted into MongoDB-compatible documents.
Delivery documents: 950
Driver documents: 170
Vehicle documents: 120
Hub documents: 8
Customer documents: 650


In [ ]:
# Aggregation 1: Summarise delivery performance by hub

hub_performance = deliveries_collection.aggregate([
    {
        "$group": {
            "_id": "$hub_id",
            "total_deliveries": {"$sum": 1},
            "delayed_deliveries": {
                "$sum": {
                    "$cond": [{"$eq": ["$delivery_status", "Delayed"]}, 1, 0]
                }
            },
            "failed_deliveries": {
                "$sum": {
                    "$cond": [{"$eq": ["$delivery_status", "Failed"]}, 1, 0]
                }
            },
            "average_rating": {"$avg": "$customer_rating_post_delivery"}
        }
    },
    {
        "$sort": {"delayed_deliveries": -1}
    }
])

print("Delivery performance by hub:")
for hub in hub_performance:
    pprint(hub)

Delivery performance by hub:
{'_id': 'H04',
 'average_rating': 3.9154761904761908,
 'delayed_deliveries': 28,
 'failed_deliveries': 16,
 'total_deliveries': 127}
{'_id': 'H06',
 'average_rating': 3.882135922330097,
 'delayed_deliveries': 27,
 'failed_deliveries': 15,
 'total_deliveries': 104}
{'_id': 'H02',
 'average_rating': 3.950952380952381,
 'delayed_deliveries': 26,
 'failed_deliveries': 10,
 'total_deliveries': 106}
{'_id': 'H01',
 'average_rating': 3.840592592592593,
 'delayed_deliveries': 26,
 'failed_deliveries': 17,
 'total_deliveries': 136}
{'_id': 'H07',
 'average_rating': 3.881858407079646,
 'delayed_deliveries': 25,
 'failed_deliveries': 14,
 'total_deliveries': 115}
{'_id': 'H05',
 'average_rating': 3.669557522123894,
 'delayed_deliveries': 25,
 'failed_deliveries': 23,
 'total_deliveries': 115}
{'_id': 'H03',
 'average_rating': 3.8958620689655175,
 'delayed_deliveries': 23,
 'failed_deliveries': 11,
 'total_deliveries': 119}
{'_id': 'H08',
 'average_rating': 3.88456,
 '

In [ ]:
# Aggregation 2: Compare average customer rating by delivery status

rating_by_status = deliveries_collection.aggregate([
    {
        "$group": {
            "_id": "$delivery_status",
            "total_deliveries": {"$sum": 1},
            "average_customer_rating": {"$avg": "$customer_rating_post_delivery"},
            "average_route_distance_km": {"$avg": "$route_distance_km"},
            "average_fuel_or_charge_cost": {"$avg": "$fuel_or_charge_cost"}
        }
    },
    {
        "$sort": {"average_customer_rating": -1}
    }
])

print("Average performance measures by delivery status:")
for status in rating_by_status:
    pprint(status)

Average performance measures by delivery status:
{'_id': 'OnTime',
 'average_customer_rating': 4.28327302631579,
 'average_fuel_or_charge_cost': 12.678051948051948,
 'average_route_distance_km': 13.776363636363635,
 'total_deliveries': 616}
{'_id': 'Delayed',
 'average_customer_rating': 3.11497461928934,
 'average_fuel_or_charge_cost': 13.13871287128713,
 'average_route_distance_km': 14.670247524752474,
 'total_deliveries': 202}
{'_id': 'Failed',
 'average_customer_rating': 3.0493129770992367,
 'average_fuel_or_charge_cost': 13.147954545454546,
 'average_route_distance_km': 13.36530303030303,
 'total_deliveries': 132}


## 9. NoSQL Relationship Handling Through Document Enrichment

Although MongoDB does not require traditional relational joins, related data can still be combined when needed. In this section, delivery records are enriched with driver and vehicle information using an aggregation pipeline with `$lookup`.

This demonstrates how MongoDB can preserve flexible document storage while still supporting analysis across multiple collections.

In [ ]:
# Use $lookup to enrich delivery records with driver and vehicle details

enriched_deliveries = deliveries_collection.aggregate([
    {
        "$lookup": {
            "from": "drivers",
            "localField": "driver_id",
            "foreignField": "driver_id",
            "as": "driver_details"
        }
    },
    {
        "$lookup": {
            "from": "vehicles",
            "localField": "vehicle_id",
            "foreignField": "vehicle_id",
            "as": "vehicle_details"
        }
    },
    {
        "$unwind": "$driver_details"
    },
    {
        "$unwind": "$vehicle_details"
    },
    {
        "$project": {
            "_id": 0,
            "delivery_id": 1,
            "delivery_status": 1,
            "driver_id": 1,
            "driver_rating": "$driver_details.driver_rating",
            "years_experience": "$driver_details.years_experience",
            "vehicle_id": 1,
            "vehicle_type": "$vehicle_details.vehicle_type",
            "maintenance_status": "$vehicle_details.maintenance_status"
        }
    },
    {
        "$limit": 5
    }
])

print("Sample enriched delivery documents:")
for delivery in enriched_deliveries:
    pprint(delivery)

Sample enriched delivery documents:
{'delivery_id': 'DL00001',
 'delivery_status': 'Failed',
 'driver_id': 'D004',
 'driver_rating': 4.75,
 'maintenance_status': 'Active',
 'vehicle_id': 'V056',
 'vehicle_type': 'EV',
 'years_experience': 13}
{'delivery_id': 'DL00002',
 'delivery_status': 'OnTime',
 'driver_id': 'D138',
 'driver_rating': 4.61,
 'maintenance_status': 'Active',
 'vehicle_id': 'V007',
 'vehicle_type': 'Diesel',
 'years_experience': 11}
{'delivery_id': 'DL00003',
 'delivery_status': 'OnTime',
 'driver_id': 'D006',
 'driver_rating': 4.38,
 'maintenance_status': 'Active',
 'vehicle_id': 'V049',
 'vehicle_type': 'Diesel',
 'years_experience': 8}
{'delivery_id': 'DL00004',
 'delivery_status': 'Delayed',
 'driver_id': 'D116',
 'driver_rating': 4.19,
 'maintenance_status': 'Active',
 'vehicle_id': 'V055',
 'vehicle_type': 'Hybrid',
 'years_experience': 4}
{'delivery_id': 'DL00005',
 'delivery_status': 'OnTime',
 'driver_id': 'D108',
 'driver_rating': 4.33,
 'maintenance_status':

## 10. Example of a More Flexible Nested Document Structure

One of the main advantages of MongoDB is that related information can be embedded inside a single document when this improves access efficiency. In a logistics system, a delivery document may be stored with selected driver and vehicle details nested inside it, allowing frequently used information to be retrieved together without requiring a separate join each time.

The following example demonstrates how one enriched delivery record can be represented as a nested MongoDB-style document.

In [ ]:
# Create one example of a nested MongoDB-style delivery document

delivery = deliveries_collection.find_one(
    {"delivery_id": "DL00001"},
    {"_id": 0}
)

driver = drivers_collection.find_one(
    {"driver_id": delivery["driver_id"]},
    {"_id": 0, "driver_id": 1, "driver_rating": 1, "years_experience": 1, "employment_type": 1}
)

vehicle = vehicles_collection.find_one(
    {"vehicle_id": delivery["vehicle_id"]},
    {"_id": 0, "vehicle_id": 1, "vehicle_type": 1, "maintenance_status": 1, "telematics_version": 1}
)

nested_delivery_document = {
    "delivery_id": delivery["delivery_id"],
    "order_id": delivery["order_id"],
    "hub_id": delivery["hub_id"],
    "delivery_status": delivery["delivery_status"],
    "route_distance_km": delivery["route_distance_km"],
    "customer_rating_post_delivery": delivery["customer_rating_post_delivery"],
    "driver": driver,
    "vehicle": vehicle
}

print("Example nested delivery document:")
pprint(nested_delivery_document)

Example nested delivery document:
{'customer_rating_post_delivery': 3.07,
 'delivery_id': 'DL00001',
 'delivery_status': 'Failed',
 'driver': {'driver_id': 'D004',
            'driver_rating': 4.75,
            'employment_type': 'PartTime',
            'years_experience': 13},
 'hub_id': 'H05',
 'order_id': 'O00938',
 'route_distance_km': 17.26,
 'vehicle': {'maintenance_status': 'Active',
             'telematics_version': 'v2.2',
             'vehicle_id': 'V056',
             'vehicle_type': 'EV'}}


## 11. CRUD Operations in MongoDB

MongoDB supports full CRUD functionality: **Create, Read, Update, and Delete**. Earlier sections already demonstrated document creation through insertion and document reading through queries. The following controlled example demonstrates update and delete operations using a temporary test document so that the main analytical dataset remains unchanged.

In [4]:
# Reconnect to MongoDB Atlas and select the deliveries collection if needed

from pymongo import MongoClient
from getpass import getpass
from urllib.parse import quote_plus
from pprint import pprint

username = "northstar_user"
password = quote_plus(getpass("Enter your MongoDB Atlas password: "))

mongo_uri = f"mongodb+srv://{username}:{password}@cluster0.uifgv62.mongodb.net/?appName=Cluster0"

client = MongoClient(mongo_uri)
client.admin.command("ping")

db = client["northstar_logistics"]
deliveries_collection = db["deliveries"]

print("Successfully reconnected to MongoDB Atlas.")
print("Deliveries collection selected.")

Enter your MongoDB Atlas password: ··········
Successfully reconnected to MongoDB Atlas.
Deliveries collection selected.


In [5]:
# Demonstrate full CRUD operations using a temporary test document

test_document = {
    "delivery_id": "TEST_CRUD_001",
    "delivery_status": "Pending",
    "hub_id": "H01",
    "manual_route_override_count": 0,
    "customer_rating_post_delivery": None
}

# CREATE
deliveries_collection.insert_one(test_document)
print("CREATE: Temporary test document inserted.")
pprint(deliveries_collection.find_one({"delivery_id": "TEST_CRUD_001"}, {"_id": 0}))

# READ
read_result = deliveries_collection.find_one(
    {"delivery_id": "TEST_CRUD_001"},
    {"_id": 0}
)
print("\nREAD: Temporary test document retrieved.")
pprint(read_result)

# UPDATE
deliveries_collection.update_one(
    {"delivery_id": "TEST_CRUD_001"},
    {"$set": {"delivery_status": "Completed", "manual_route_override_count": 1}}
)
print("\nUPDATE: Temporary test document updated.")
pprint(deliveries_collection.find_one({"delivery_id": "TEST_CRUD_001"}, {"_id": 0}))

# DELETE
deliveries_collection.delete_one({"delivery_id": "TEST_CRUD_001"})
print("\nDELETE: Temporary test document removed.")
print("Remaining test documents:",
      deliveries_collection.count_documents({"delivery_id": "TEST_CRUD_001"}))

CREATE: Temporary test document inserted.
{'customer_rating_post_delivery': None,
 'delivery_id': 'TEST_CRUD_001',
 'delivery_status': 'Pending',
 'hub_id': 'H01',
 'manual_route_override_count': 0}

READ: Temporary test document retrieved.
{'customer_rating_post_delivery': None,
 'delivery_id': 'TEST_CRUD_001',
 'delivery_status': 'Pending',
 'hub_id': 'H01',
 'manual_route_override_count': 0}

UPDATE: Temporary test document updated.
{'customer_rating_post_delivery': None,
 'delivery_id': 'TEST_CRUD_001',
 'delivery_status': 'Completed',
 'hub_id': 'H01',
 'manual_route_override_count': 1}

DELETE: Temporary test document removed.
Remaining test documents: 0


The CRUD demonstration confirms that MongoDB Atlas was not only used for data insertion and querying, but also for controlled document modification and removal. A temporary document was used so that the main operational delivery dataset remained unchanged after testing.

## 12. Key Findings from the MongoDB Analysis

The MongoDB development process successfully created a cloud-hosted NoSQL database for NorthStar using separate collections for deliveries, drivers, vehicles, hubs, and customers.

The analysis produced several useful operational findings:

- **616 deliveries were completed on time**, while **202 were delayed** and **132 failed**.
- **Hub H04** recorded the highest number of delayed deliveries with **28 delays**.
- **Hub H08** recorded the highest number of failed deliveries with **26 failures**.
- On-time deliveries achieved a much higher average customer rating of approximately **4.28**, compared with **3.11** for delayed deliveries and **3.05** for failed deliveries.
- MongoDB queries successfully identified deliveries with missing proof of completion and deliveries with unusually high manual route override counts.
- The `$lookup` pipeline demonstrated that related information from multiple collections can be combined when required.
- The nested document example showed how MongoDB can store selected related details together inside one flexible document, which is useful for operational systems that frequently retrieve connected delivery information.

## 13. Conclusion

This notebook demonstrated the development of a MongoDB Atlas NoSQL database for the NorthStar Urban Mobility and Logistics case study using Python and PyMongo.

The work showed how structured CSV datasets can be transformed into MongoDB-compatible documents, inserted into cloud-hosted collections, and queried using both standard document queries and aggregation pipelines. The notebook also demonstrated how MongoDB can support flexible document structures through nested documents and can still combine related collections using `$lookup` when cross-collection analysis is required.

Overall, MongoDB provides NorthStar with a scalable and flexible database model that is well suited to operational logistics data, especially where delivery records, vehicle information, driver details, and evolving business requirements need to be handled efficiently.